### DISCUSSION TITLE
Navigation display progress with elevation diagram

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#412] [Q&A] [answered]
A graphical representation of the navigation progress and elevation ahead on the route.

Elevation graph is plotted on two axes: X-axis is the distance of the route and Y-axis is the elevation.
The appearance of the graph uses the colors and transparency selected in the app settings.

Gestures with the graph (similar to map):

- Zoom in/out. The graph centers on the current location and shows the route up to a distance ahead. Use two-finger movement (or double tap) to zoom in / out for a more detailed view of elevation changes. Move the graph forward on the route and backward to the current location.

- The current location moves on the graph and the label shows the elevation and distance traveled. The graph auto centers on the current location after a while.

More route details may be added in the future.

<img src="https://github.com/devemux86/cruiser/assets/3484020/75df8ff5-2b6d-477f-9667-151bbdd508ac" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/d178d198-914e-491c-ad9b-2785c8d6dfc0" width="50%"/>
![Landscape_1](https://github.com/devemux86/cruiser/assets/3484020/33f116b6-ddcc-4b4f-82d3-76e1208f1593)
![Landscape_2](https://github.com/devemux86/cruiser/assets/3484020/08b94990-e717-4897-ac08-77aa70a73446)

**Comment by Ciutnik** [comment]
Oh man, it looks really great! Big feature 👍

Edit: Those using landscape mode may have a problem. Unfortunately, this also applies to me...

**Reply by devemux86** [reply]
@Ciutnik I added above some landscape images.

The graph panel supports the same appearance settings, so it can have transparency.

I don't know if it makes sense to have it vertical on the side in landscape mode.

**Reply by Ciutnik** [reply]
Thanks for the horizontal pictures. Contrary to fears, they don't look too bad.

> I don't know if it makes sense to have it vertical on the side in landscape mode.

I was wondering about that too, but either the graph would be terribly narrow or rotated by 90 degrees, which in the case of a height graph would be at least strange, and certainly not very readable 🙂

But maybe a narrow strip with the color of the section depending on the slope (and an indication of our position) would be nice? This could replace the wide bar of the elevation chart. I think that slope is important information not only for cyclists 🤔 Maybe even more important than height? These are just my thoughts...

**Reply by airald** [reply]
I like a lot the landscape images. If this space is needed to show the route, horiz axis legend could be placed over the graphic and just draw the altitude profile, maybe with a high width and no background (or high transparency). Something like that:

![profile](https://github.com/devemux86/cruiser/assets/20151719/e01d82d5-ca4e-4e02-a476-22e88625fe07)

**Reply by DolomitesTom** [reply]
transparent would be fine. The route should always be in the foreground.

**Reply by bios4** [reply]
> Something like that

Nice suggestion! 👍

**Comment by devemux86** [comment]
The first implementation is available for testing in:
- https://github.com/devemux86/cruiser/discussions/414

It follows the existing color / transparency settings.
You can also try it with navigation simulation to see what improvements are needed.

**Comment by devemux86** [comment]
Playing with the graph appearance:

![Screenshot_1](https://github.com/devemux86/cruiser/assets/3484020/e0d8bdc2-7e11-4703-8059-7abfc8886df0)
![Screenshot_2](https://github.com/devemux86/cruiser/assets/3484020/9cdb446e-0c1a-47ec-aeed-0fd84d4b922b)

**Comment by DolomitesTom** [comment]
How do I get the text above the graph instead of below the graph?

![Screenshot_20240104-174251_Cruiser](https://github.com/devemux86/cruiser/assets/121371629/0be03a21-75a1-4ce8-a13b-82765dda0715)

**Reply by devemux86** [reply]
@DolomitesTom need a new version. :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
Ok. And the km bar in the graph jumps from the left to the right, to middle and back to left.... and also disappears sometimes. This is a bit irritating. Shouldn't it stay at one position and just the km scale should move?

**Reply by devemux86** [reply]
@DolomitesTom

The location bar on the graph is usually at the left, so we can see more route ahead of us on the right.
By default the graph is zoomed so several km are ahead on the right. We can zoom in / out the graph.

It can disappear if there is no GPS location or we go off route. Were you tapping on the graph?
Tap on the graph moves the location bar temporarily to indicate the tapped elevation / distance.
After some seconds it returns to the GPS location.

**Reply by DolomitesTom** [reply]
Thanks for explanation, not tried yet under live conditions.

**Comment by Ciutnik** [comment]
I'm just thinking... Could panels placed centrally (chart, street names, next waypoint) have separate colors settings? For better readability, I have turned off the transparency of the panels (solid black background). But if I want to see the graph, I limit the map area very much, especially in landscape mode. So I would love to make it e.g. semi transparent on a light background.

**Reply by devemux86** [reply]
@Ciutnik do you mean only the background color?

We could probably have a setting for the graph background, as it is a large panel.

**Reply by Ciutnik** [reply]
Won't changing the background make the graph disappear under certain conditions? There would probably have to be some relationship between the background color and the fill color of the chart.

When writing above about "centrally placed panels", I didn't think that in vertical mode their width includes the side panels, so they become upper panels rather than central panels. In this case, their different color than the side panels would look bad.

**Comment by umpfmpf** [comment]
For me, the graph jumps in some places every now and then. 

https://github.com/devemux86/cruiser/assets/105640513/0c8fd8b5-590f-45c8-8049-271b34514b1e

He simply moves for a short moment and then returns to the normal position. Sometimes he jumps very far away and not only sees these few meters as in the video.

**Reply by devemux86** [reply]
@umpfmpf thanks for the video.

The location bar stays at a stable position (before 108 km), the graphs moves.
Maybe it is better to set the location bar at X-axis center to avoid such issues?

**Reply by bios4** [reply]
> Maybe it is better to set the location bar at X-axis center to avoid such issues?

I would highly appreciate if the location bar would stay at a FIXED POSITION (very gladly in the position shown at the beginning of the video, approx. 1 cm from the left border of the chart) and only the underlying altitude chart and its km-scale (x-axis) would move forward...!

EDIT: location bar further left instead of x-axis center, because most of the time it is more important what is ahead, instead of what is behind me!

Additionally, I agree width airald who stated "_horiz axis legend could be placed over the graphic and just draw the altitude profile, maybe with a high width and no background (or high transparency)._" - to be seen [here](https://github.com/devemux86/cruiser/discussions/412#discussioncomment-7993849).

**Reply by umpfmpf** [reply]
The position on the left is already good, so the following route can be seen. 
Only that it jumps in some places (reproducible) is strange.

**Reply by devemux86** [reply]
And the (currently top) label with current elevation / distance, in what position?

Or we do not need it?

**Reply by umpfmpf** [reply]
Not really. The height is already on the right side.

**Reply by devemux86** [reply]
They are not exactly the same:
- The graph displays the elevation of the route (provided by the route)

- The (right) panel displays the altitude of location (provided by the GPS) and it also works when being off route
(also we can adjust it for mean see level in location settings)

Although they may have the same value when we are on route.

**Reply by umpfmpf** [reply]
So it is actually not so important for me when we are on the route. The font in the right panel is larger and I can read it better while driving. 
And if the space can be used better, the graph is also without clear.

**Reply by devemux86** [reply]
Indeed, better keep things simple and more visible.

**Comment by devemux86** [comment]
> I agree width airald who stated "horiz axis legend could be placed over the graphic and just draw the altitude profile, maybe with a high width and no background (or high transparency)."

Graph filled and not, app panel transparency 50% (outline optional).

<img src="https://github.com/devemux86/cruiser/assets/3484020/c13c811f-a921-445f-a3fb-b2d2717b8633" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/d1d885e4-b244-4b3c-9109-d2ef08dee2be" width="50%"/>
<img src="https://github.com/devemux86/cruiser/assets/3484020/acc294b9-5810-4df8-8b42-aeb5b4f2d160" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/19326212-6d71-44c1-af82-ada0b3bda20e" width="50%"/>

**Comment by GuG1050** [comment]
Nice Feature but wll the graph be selectable,, can it be switched off?

**Reply by devemux86** [reply]
@GuG1050 of course, like all Cruiser features. :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
When is average speed with and without pause introduced in the track details?

**Reply by devemux86** [reply]
@DolomitesTom we will see, sometime in the future. :slightly_smiling_face:

**Reply by devemux86** [reply]
We can first work with what is immediately available, such as elevation.
Other values have to be processed separately, much more work.

**Comment by umpfmpf** [comment]
For me, the distance font runs in high-altitude font and makes it illegible.
![Screenshot_20240106_180817_Cruiser](https://github.com/devemux86/cruiser/assets/105640513/81576264-2a38-4888-a63b-9eb309e94d1e)
![Screenshot_20240106_180627_Cruiser](https://github.com/devemux86/cruiser/assets/105640513/50c9ffce-5c95-4573-8b64-2fcb07fcd91d)

**Reply by devemux86** [reply]
@umpfmpf if you enable the outline setting (default on), such texts should be readable (like in images [above](https://github.com/devemux86/cruiser/discussions/412#discussioncomment-8025452)).

Or we could place the X-axis labels on top of the axis.

**Comment by ramondevic** [comment]
In GraphHopper mode it doesn't show, it only shows with BRouter routing mode, right?

I always use the GraphHopper to be able to see the name of the current street or road and the name of the next one.
If the streets and the graph were shown, we would run out of space to see the map, so perhaps you could use GraphHopper to show/hide the graph with one click.

Could the height of the graph consume the same space as the three lines of destination, next street and current street shown in graphhopper?

I would also like the fields in the right panel to be located in the exact same place at the same screen coordinate so that some automations I use don't move.

![Screenshot_20240106-182802](https://github.com/devemux86/cruiser/assets/122691401/4a854adb-e1fe-4258-8d22-5040f45c1345) ![Screenshot_20240106-182929](https://github.com/devemux86/cruiser/assets/122691401/29e509b1-2b13-47d9-a160-8d91a894ed74)

**Reply by devemux86** [reply]
@ramondevic 

> In GraphHopper mode it doesn't show, it only shows with BRouter routing mode, right?

GraphHopper (offline) routing currently does not contain elevation data.

> Could the height of the graph consume the same space as the three lines of destination, next street and current street shown in graphhopper?

It should already have such height, compare the 2 images.
(the left image has an extra panel on the top)

> I would also like the fields in the right panel to be located in the exact same place at the same screen coordinate so that some automations I use don't move.

In portrait mode, it depends on the top panels, the right panels are located below the top panels.

**Comment by umpfmpf** [comment]
On top of the axis would perhaps better. With outline it also looks not so good. 
![Screenshot_20240106_192440_Cruiser](https://github.com/devemux86/cruiser/assets/105640513/e01e5e59-359f-40a9-885d-1830c708b3a0)
![Screenshot_20240106_191417_Cruiser](https://github.com/devemux86/cruiser/assets/105640513/2b93a3e5-4af9-4631-8a37-935e11e52750)

**Comment by devemux86** [comment]
Should the labels be at the top (above the axis) or at the bottom (below the axis)?
(at the bottom we gain some pixels for the graph)

<img src="https://github.com/devemux86/cruiser/assets/3484020/8e561b73-97b5-4397-836b-461501c46dd6" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/bd77ee6e-7e08-4a7c-833c-3efed31f4c75" width="50%"/>

**Comment by devemux86** [comment]
There is also an alternative appearance with no borders between the panels (image 1).
It seems more "continuous", only drawback is that we have no borders even in absolute transparency.

Do you prefer borders or not?

<img src="https://github.com/devemux86/cruiser/assets/3484020/dd4058d9-43ab-4ba4-a26b-d6469d80e535" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/8e561b73-97b5-4397-836b-461501c46dd6" width="50%"/>

**Reply by umpfmpf** [reply]
For me it looks better with borders between the panels and the letters on top of the graph.

**Reply by Ciutnik** [reply]
> For me it looks better with borders between the panels and the letters on top of the graph.

I fully agree with @umpfmpf

**Reply by airald** [reply]
My opinion is the same than @umpfmpf and @Ciutnik

**Comment by devemux86** [comment]
Waypoints will be visible in the navigation progress graph:

<img src="https://github.com/devemux86/cruiser/assets/3484020/6ccd127b-dc08-41e4-aa67-1c8973bb1f1f" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/ae8b2bd6-d2a9-49fc-9d4b-2c9abd552242" width="50%"/>

**Comment by devemux86** [comment]
Could this maybe be changed to allow de-/activating the graph also during a running navigation?

_Originally posted by @bios4 in https://github.com/devemux86/cruiser/discussions/411#discussioncomment-8181349._

**Reply by devemux86** [reply]
We could add a button in navigation options (long press on direction icon) to turn on / off the graph during navigation?

**Reply by devemux86** [reply]
Another idea is to open a menu with buttons where we can turn on / off all panels during navigation:
graph, top panels, right and left panels, etc.

**Reply by devemux86** [reply]
<img src="https://github.com/devemux86/cruiser/assets/3484020/ddd320b9-eecd-4b7e-9fc7-5c4511c5a99e" width="50%"/>

**Comment by devemux86** [comment, marked as answer]
Navigation display progress with elevation diagram is available in [Cruiser 3.0.38](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
